### PR TITLE
php 8.1.0

### PR DESCRIPTION
**Pull request by ruudk**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by SMillerDev** [comment]
Can you try removing:
- The patch on line 68 and at the end of the file
- The sed override on line 107
- The manual command on line 320

And see if it works without those?

**Comment by carlocab** [comment]
We should also try to change
```ruby
depends_on "libffi"
```
to
```ruby
uses_from_macos "libffi", since: :catalina
```

File: Formula/php.rb [hunk 1]
```diff
@@ -61,12 +60,7 @@ class Php < Formula
   uses_from_macos "libxml2"
   uses_from_macos "libxslt"
   uses_from_macos "zlib"
-
-  on_macos do
-    # PHP build system incorrectly links system libraries
-    # see https://github.com/php/php-src/pull/3472
-    patch :DATA
```
Comment (carlocab) [inline Formula/php.rb:68]: The patch is still needed. Or, at least, some of it is.
```
configure: WARNING: unrecognized options: --enable-phpdbg-webhelper, --with-xmlrpc, --with-os-sdkpath
```

Comment (ruudk) [reply Formula/php.rb:68]: If we remove the patch, shouldn't we remove the `--with-os-sdkpath` option too?

**Comment by derrabus** [comment]
PHP CS Fixer will refuse to run on PHP 8.1 which will break the CI. How do we tackle this?

File: Formula/php.rb (hunk 1 above)
Comment (carlocab) [reply Formula/php.rb:68]: Not being able to pass `--with-os-sdkpath` is probably breaking the build though.

Comment (ruudk) [reply Formula/php.rb]: I see, I added the patch back.

**Comment by ruudk** [comment]
@derrabus What about running it with `PHP_CS_FIXER_IGNORE_ENV=1`?

File: Formula/php.rb [hunk 2]
```diff
@@ -61,6 +60,7 @@ class Php < Formula
   uses_from_macos "libxml2"
   uses_from_macos "libxslt"
   uses_from_macos "zlib"
+  uses_from_macos "libffi", since: :catalina
```
Comment (carlocab) [inline Formula/php.rb:63]: This probably needs to be after `libedit` but before `libx*`. The audit will complain if needed, though.

Comment (ruudk) [reply Formula/php.rb]: Fixed.

**Comment by derrabus** [comment]
> What about running it with `PHP_CS_FIXER_IGNORE_ENV=1`?

Just for the tests? Might be worth a try. But that would just fix the tests, the problem that the user would get an error when runnin PHP CS Fixer will remain until a compatible release has been tagged.

`PHP_CS_FIXER_IGNORE_ENV` is somewhat a "I know what I'm doing" flag. I don't think we should set it globally. Besides, I don't expect the `php-cs-fixer@2` formula to receive compat fixes for PHP 8.1.

Do we have a way to choose the PHP version a PHP tool should run on?

**Comment by ruudk** [comment]
@derrabus I meant for the test indeed. Not sure how to handle at normal runtime. 

@carlocab Would it be possible to label the PR with `CI-no-fail-fast` so that we can see if the other builds do finish? It might be related to ARM?

**Comment by SMillerDev** [comment]
> PHP CS Fixer will refuse to run on PHP 8.1 which will break the CI. How do we tackle this?

Make it depend on an older version

**Comment by ruudk** [comment]
> Make it depend on an older version

Should we copy the old php formula to `php@8.0` so that we can do it in this same PR?

**Comment by derrabus** [comment]
> Make it depend on an older version

That won't be enough because the PHAR's shebang would point to the globally linked PHP version (which is 8.1 after this PR). Am I thinking too complicated? 🙈

**Comment by derrabus** [comment]
I'm trying to pin PHP CS Fixer to a specific PHP version: #89975

**Comment by ruudk** [comment]
Would it be possible to label the PR with CI-no-fail-fast so that we can see if the other builds do finish? It might be related to ARM?

**Comment by ruudk** [comment]
I'm getting this error:
```
  php:
    * 'revision 1' should be removed
  Error: 1 problem in 1 formula detected
  Error: 'revision 1' should be removed
```

not sure where this comes from.

**Comment by derrabus** [comment]
@ruudk The PHP formula got a revision bump since you've branched out. Rebase your branch on master and remove the revision.

**Comment by carlocab** [comment]
I think your branch is out of date. BrewTestBot uses the version of the formula you get after rebasing your commits on master.

**Comment by ruudk** [comment]
Added `PHP_CS_FIXER_IGNORE_ENV=1` for the php-cs-fixer tests in ddfcb12 to let them pass.

The ARM64 builds seem to break on `  make: *** [ext/opcache/jit/zend_jit.lo] Error 1`

Anybody an idea how to solve that?

**Comment by ruudk** [comment]
Full error:
```
/bin/sh /private/tmp/php-20211126-1297-1394jiu/php-8.1.0/libtool --silent --preserve-dup-deps --tag CC --mode=compile clang -Iext/opcache/ -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0/ext/opcache/ -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0/include -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0/main -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0 -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0/ext/date/lib -I/opt/homebrew/Cellar/openssl@1.1/1.1.1l_1/include -I/opt/homebrew/Cellar/pcre2/10.39/include -I/opt/homebrew/Cellar/sqlite/3.36.0/include -I/opt/homebrew/Cellar/curl/7.80.0_1/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/ffi -I/opt/homebrew/Cellar/gd/2.3.3_1/include -I/opt/homebrew/opt/gettext/include -I/opt/homebrew/opt/gmp/include -I/opt/homebrew/Cellar/icu4c/69.1/include -I/opt/homebrew/opt/openldap/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sasl -I/opt/homebrew/Cellar/oniguruma/6.9.7.1/include -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0/ext/mbstring/libmbfl -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0/ext/mbstring/libmbfl/mbfl -I/opt/homebrew/Cellar/unixodbc/2.3.9_1/include -I/opt/homebrew/opt/freetds/include -I/opt/homebrew/opt/libpq/include -I/opt/homebrew/opt/aspell/include/pspell -I/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/editline -I/opt/homebrew/Cellar/libsodium/1.0.18_1/include -I/opt/homebrew/Cellar/argon2/20190702_1/include -I/opt/homebrew/opt/tidy-html5/include -I/opt/homebrew/Cellar/libzip/1.8.0_1/include -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0/TSRM -I/private/tmp/php-20211126-1297-1394jiu/php-8.1.0/Zend    -fno-common -Wstrict-prototypes -Wall -Wextra -Wno-strict-aliasing -Wno-unused-parameter -Wno-sign-compare -g -O2 -fvisibility=hidden -DZEND_SIGNALS   -prefer-pic -Wno-implicit-fallthrough -DZEND_ENABLE_STATIC_TSRMLS_CACHE=1 -c /private/tmp/php-20211126-1297-1394jiu/php-8.1.0/ext/opcache/jit/zend_jit.c -o ext/opcache/jit/zend_jit.lo  -MMD -MF ext/opcache/jit/zend_jit.dep -MT ext/opcache/jit/zend_jit.lo
clang: error: unable to execute command: Segmentation fault: 11
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: arm64-apple-darwin21.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /private/tmp/zend_jit-eed91f.c
clang: note: diagnostic msg: /private/tmp/zend_jit-eed91f.sh
clang: note: diagnostic msg: Crash backtrace is located in
clang: note: diagnostic msg: /private/tmp/php-20211126-1297-1394jiu/php-8.1.0/.brew_home/Library/Logs/DiagnosticReports/clang_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
clang: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
clang: note: diagnostic msg: 

********************
make: *** [ext/opcache/jit/zend_jit.lo] Error 1
```

**Comment by derrabus** [comment]
I can reproduce the build error locally. The build passes if I revert the two lines related to libffi.

**Comment by ruudk** [comment]
@derrabus Still fails :(

**Comment by derrabus** [comment]
It might have been a coincidence that the build passed after reverting libffi. I've run 10 builds with your branch now. Four of them failed, six passed. 😕

**Comment by carlocab** [comment]
Yea, it's almost certainly not related to `libffi`. Apple's version is compatible with the one we ship (plus macOS-specific improvements by Apple).

**Comment by aurimasniekis** [comment]
Any progress on this?

**Comment by ruudk** [comment]
See conversation. It segfaults on Monterey.

**Comment by aurimasniekis** [comment]
I have been using this for a long time with 8.1 no segfaults or anything on several systems with different macOS versions

https://github.com/shivammathur/homebrew-php/blob/master/Formula/php.rb

**Comment by ruudk** [comment]
I also use that version locally without issues. So I diffed their formula with ours and I couldn't spot differences that might cause this. Does their formula also build on Monterey on CI or does it only build for older macOS versions on their CI? Maybe its a CI related issue?

**Comment by xorik** [comment]
I can help with testing. I have macos 12 on arm64

**Comment by derrabus** [comment]
> Maybe its a CI related issue?

No, it is reproducible outside of the CI. On my machine, the build with this fails about 40% of the time, but when it passes, I get a perfectly fine binary.

**Comment by ruudk** [comment]
@derrabus can you maybe share how to build this locally?

**Comment by SMillerDev** [comment]
You can run `brew reinstall -s php` with the branch checked out.

**Comment by xorik** [comment]
To install on localhost:
```
git clone git@github.com:ruudk/homebrew-core.git --depth=1 -b bump-php-8.1.0 /tmp/brew-php8.1
brew install --build-from-source  /tmp/brew-php8.1/Formula/php.rb
rm -rf /tmp/brew-php8.1
```

P.S. I think issue is that for some reason the install command returns non-zero code, but it was successfully installed:
Screen Shot 2021-11-27 at 10 50 52

**Comment by ruudk** [comment]
You can reproduce the segmentation fault locally with this command:
```bash
COUNTER=1; while brew reinstall --formula --build-from-source Formula/php.rb; do COUNTER=$[$COUNTER +1]; echo "Iteration $COUNTER"; done;
```

For me, it already fails on the second iteration.

**Comment by SMillerDev** [comment]
Is that on ARM? 

Either way this should be reported upstream.

**Comment by ruudk** [comment]
Yes, it's on macOS 12 on M1 Pro.

**Comment by ruudk** [comment]
Created a bug report here https://bugs.php.net/bug.php?id=81666

**Comment by ruudk** [comment]
It's green, but that's just a lucky shot.

**Comment by SMillerDev** [comment]
Lucky shot or not, if the binary works it works. @carlocab how do you feel about merging this?

**Comment by carlocab** [comment]
Fine, but the second commit needs to be split to get rid of the `automerge-skip` label.

**Comment by ruudk** [comment]
If I remove this, the PHP CS Fixer test will fail. Should I first submit that as a separate PR?

**Comment by ruudk** [comment]
Created a separate PR for the PHP CS Fixer fix: https://github.com/Homebrew/homebrew-core/pull/90055

**Comment by carlocab** [comment]
Sorry, to clarify: the second commit needs to be split into two in this PR, and not moved to a separate PR.

This is from our commit style guide. Here's a summary:

Please use the preferred commit-message style for homebrew/core. We put the name of the formula first in commit-message headings.

For new formulae:

> At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.

For existing formulae:

> The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.

Refer to the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) for more details. Also, when making further changes to your pull request, use the following guidelines to make sure that @BrewTestBot can merge your commits:

> - One formula per commit; one commit per formula.
> - Keep merge commits out of the pull request.

**Comment by ruudk** [comment]
I tried @SMillerDev's suggestion in https://github.com/Homebrew/homebrew-core/pull/90055#issuecomment-980642071 to change the php dependency of php-cs-fixer but that leads to the error:
```
  Error: ==> Testing php-cs-fixer
  ==> /opt/homebrew/Cellar/php-cs-fixer/3.3.2/bin/php-cs-fixer fix test.php
  env: php: No such file or directory
  Error: php-cs-fixer: failed
```

I would suggest to keep the ENV hack for now, and then merge https://github.com/Homebrew/homebrew-core/pull/89978 afterwards. 

Should we also create the new php@8.0 formula (copy of the old php.rb) in this PR or should that be done separately?

**Comment by carlocab** [comment]
You didn't actually split your second commit into two. See:

> - One formula per commit; one commit per formula.

I've fixed that for you.

**Comment by ruudk** [comment]
It's green now, anything holding back merging this? The `automerge-skip` label is still there, does that have to be removed manually?

**Comment by ruudk** [comment]
The [bug report](https://bugs.php.net/bug.php?id=81666&edit=2) received a question:
> Does it happen also with other compilers than the Xcode's one ? gcc or clang from homebrew for ex.

Is it possible to try that? Did we ever have to do that to build a PHP formula?

**Comment by aadmathijssen** [comment]
Hi @ruudk, thanks for picking this up! Regarding your question on whether to copy the current formula to `php@8.0`, I think this should definitely be done. And probably in this PR as well, otherwise we might run into the same issue as with the PHP 7.4 to 8.0 upgrade (see issue #65785).

**Comment by SMillerDev** [comment]
> Is it possible to try that? Did we ever have to do that to build a PHP formula?

We never did, but you can add a `depends_on "llvm"` locally and see if you get the same.

**Comment by SMillerDev** [comment]
> I think this should definitely be done. And probably in this PR as well, otherwise we might run into the same issue as with the PHP 7.4 to 8.0 upgrade (see issue #65785).

It shouldn't be done in this PR. This pull request is only about updating the existing formula. Adding php@8.0 would be a new formula.

**Comment by ruudk** [comment]
It would be great if Homebrew would have a bump-formula-major that took care of both things in 1 PR :)

**Comment by ruudk** [comment]
> We never did, but you can add a `depends_on "llvm"` locally and see if you get the same.

That doesn't seem to make any difference.

How can I verify that it did change the compiler?

When I open `/Users/Ruud/Library/Logs/Homebrew/php/03.make` I see:
```
Clang: 13.0.0 build 1300
Xcode: N/A
Rosetta 2: false
HOMEBREW_CC: clang
HOMEBREW_CXX: clang++
```

**Comment by carlocab** [comment]
You need to add
```ruby
fails_with :clang
fails_with :gcc
```
to use Homebrew Clang.

**Review by carlocab — APPROVED** [review]
This should be fine now. Thanks @ruudk!

**Comment by ruudk** [comment]
Thanks for merging!

@aadmathijssen I'll add the php@8.0 formula here https://github.com/Homebrew/homebrew-core/pull/90085

**Pull request closed without merging**